src/Framework/Framework/Compilation/ControlTree/BindingExtensionParameter.cs

```diff
@@ -211,6 +211,52 @@ public ViewModuleAnnotation(string id, bool isMarkupControl)
         }
     }
 
+    public class DotnetExtensionParameter : BindingExtensionParameter
+    {
+        public string Id { get; }
+        public bool IsMarkupControl { get; }
+        public ITypeDescriptor Type { get; }
+
+        public DotnetExtensionParameter(string id, bool isMarkupControl, ITypeDescriptor type) : base("_dotnet", type, true)
+        {
+            this.Id = id;
+            this.IsMarkupControl = isMarkupControl;
+            this.Type = type;
+        }
+
+        public override Expression GetServerEquivalent(Expression controlParameter)
+        {
+            var type = ResolvedTypeDescriptor.ToSystemType(this.Type);
+            var constructors = type.GetConstructors(BindingFlags.Public);
+            if (constructors.Length != 1 || constructors[0].GetParameters().Length != 1)
+            {
+                throw new DotvvmCompilationException($"The type {type} referenced in the @dotnet directive must have exactly one public constructor with one parameter of IViewModuleContext!");
+            }
+            // TODO: check parameter type
+            return Expression.New(constructors[0], Expression.Constant(null, typeof(object)))
```

Comment:
```suggestion
            return Expression.New(constructors[0], Expression.Constant(null, typeof(IViewModuleContext)))
```

It won't compile with `object`.

However, I'm thinking if we want to allow invoking the methods server-side. I guess it makes sense, but passing null into the constructor will just make it fail on the unexpected null reference. Maybe only allow it if the type also has parameterless constructor?